R/call.R

```diff
@@ -64,13 +235,43 @@ call_modify <- function(call, new_args, env = parent.frame()) {
   call
 }
 
-#' @rdname call_modify
+get_call <- function(call) {
+  if (is_frame(call)) {
+    call$expr
+  } else if (is_formula(call)) {
+    f_rhs(call)
+  } else {
+    call
+  }
+}
+
+#' Function name of a call
+#'
+#' @inheritParams call_standardise
+#' @return A string with the function name.
 #' @export
-call_standardise <- function(call, env = parent.frame()) {
-  stopifnot(is_call(call))
+#' @examples
+#' # Extract the function name from quoted calls:
+#' call_fn_name(~foo(bar))
+#' call_fn_name(quote(foo(bar)))
+#'
+#' # The calling expression is used as default:
+#' foo <- function(bar) call_fn_name()
+#' foo(bar)
+#'
+#' # Namespaced calls are correctly handled:
+#' call_fn_name(~base::matrix(baz))
+#' call_fn_name(~foo$bar(baz))
+call_fn_name <- function(call = NULL) {
+  call <- call %||% call_frame(2)
+  call <- get_call(call)
+  stopifnot(is.call(call))
+  fn <- call[[1]]
 
-  f <- eval(call[[1]], env)
-  if (is.primitive(f)) return(call)
+  # Handle namespaced calls: foo::bar(), foo@bar(), foo$bar()
```

Comment:
Yeah, that's what I was thinking. It seems like you should explicitly test that `f` has length 3, and the first element is one of `@`, `$`, `::`, `:::` or `[[`.